ethereum/p2p/src/main/java/tech/pegasys/pantheon/ethereum/p2p/upnp/UpnpNatManager.java

```diff
@@ -0,0 +1,487 @@
+    // this(new UpnpServiceImpl(new DefaultUpnpServiceConfiguration()));
+
+    // Workaround for an issue in the jupnp library: the ExecutorService used misconfigures
+    // its ThreadPoolExecutor, causing it to only launch a single thread. This prevents any work
+    // from getting done (effectively a deadlock). The issue is fixed here:
+    //   https://github.com/jupnp/jupnp/pull/117
+    // However, this fix has not made it into any releases yet.
+    // TODO: once a new release is available, remove this @Override
+    this(new UpnpServiceImpl(new PantheonUpnpServiceConfiguration()));
+  }
+
+  /**
+   * Constructor
+   *
+   * @param service is the desired instance of UpnpService.
+   */
+  public UpnpNatManager(final UpnpService service) {
+    upnpService = service;
+
+    // prime our recognizedServices map so we can use its key-set later
+    recognizedServices = new HashMap<>();
+    recognizedServices.put(SERVICE_TYPE_WAN_IP_CONNECTION, null);
+
+    // registry listener to observe new devices and look for specific services
+    registryListener =
+        new DefaultRegistryListener() {
+          @Override
+          public void remoteDeviceAdded(final Registry registry, final RemoteDevice device) {
+            LOG.debug("UPnP Device discovered: " + device.getDetails().getFriendlyName());
+            inspectDeviceRecursive(device, recognizedServices.keySet());
+          }
+        };
+  }
+
+  /**
+   * Start the manager. Must not be in started state.
+   *
+   * @throws IllegalStateException if already started.
+   */
+  public synchronized void start() {
+    if (started) {
+      throw new IllegalStateException("Cannot start already-started service");
+    }
+
+    LOG.info("starting upnp service...");
+    upnpService.startup();
+    upnpService.getRegistry().addListener(registryListener);
+
+    // TODO: does jupnp do this automatically?
+    upnpService.getControlPoint().search(new STAllHeader());
+
+    initiateExternalIpQuery();
+
+    started = true;
+  }
+
+  /**
+   * Stop the manager. Must not be in stopped state.
+   *
+   * @throws IllegalStateException if stopped.
+   */
+  public synchronized void stop() {
+    if (!started) {
+      throw new IllegalStateException("Cannot stop already-stopped service");
+    }
+    upnpService.getRegistry().removeListener(registryListener);
+    upnpService.shutdown();
+
+    started = false;
+  }
+
+  /**
+   * Returns the first of the discovered services of the given type, if any.
+   *
+   * @param type is the type descriptor of the desired service
+   * @return the first instance of the given type, or null if none
+   */
+  private synchronized RemoteService getService(final String type) {
+    return recognizedServices.get(type);
+  }
+
+  /**
+   * Get the discovered WANIPConnection service, if any.
+   *
+   * @return the WANIPConnection Service if we have found it, or null.
+   */
+  public synchronized RemoteService getWANIPConnectionService() {
```

Comment:
Is this just exposed for testing?  If so, I'd suggest cutting this and making `getService` package-private (and marking it `@VisibleForTesting`) so that tests can access it.

Reply:
1. Removed note about class not being threadsafe, it should be pretty threadsafe at this point.
2. Yes, synchronization mostly relates to `recognizedServices` and `started`, maybe others (`externalIpQueryFuture`, for example)
3. `getWANIPConnectionService` updated as suggested